### DISCUSSION TITLE
How to properly narrowing down resource or accessor type in Solid.js

### DISCUSSION THREAD

**Discussion by ShueiYang** [solidjs/solid#1806] [Q&A] [answered]
Hello , I start to play with Solid.js and also new with Typescript and I have hard time to figure out how to narrow properly the type in solid... a simple example  with the code below after fetching the ressource: 

```js
const App: Component = () => {
  
  const [ pageData ] = createResource<DeliveryData>(getPageData, {initialValue: undefined});


  return (
    <Show 
      when={!pageData.loading && pageData()}
      fallback={<Loader />}
    >
    <>
      <Header datas={pageData()}/>
      <MainContent datas={pageData()}/>
      <Footer />
    </>
    </Show>
  );
};

export default App;
```
Typescript is complaining about the fact <DeliveryData | undefined> cannot be assert into....etc  I try to narrow down to the DeliveryData only by using conditional rendering with && operator, ternacy operator or typeof on the Show or in the component but it didn't work,  Typescript said datas pass to the child can still be undefined... the only way is to use assertion as DeliveryData on datas props.. but this is not really checking so what did I miss or doing wrong ?

EDIT ok I just find out that we can also use callback in the show like this
```js
    <Show 
      when={!pageData.loading && pageData()}
      fallback={<Loader />}
    >
      { pageData => 
        <>
          <Header datas={pageData()}/>
          <MainContent datas={pageData()}/>
          <Footer />
        </> 
      }
    </Show>
```

but what will happen if we store the signal outside and i want to call it from the "store"  ?

**Comment by snnsnn** [comment, marked as answer]
A resource's data is `undefined` initially and resoves to its actual value once the promise is resolved. Typescript can not see the conditional statement that comes from a component, so you need use none null assertion `!`:

```tsx
<Switch>
  <Match when={data.state === 'pending'}>Pending</Match>
  <Match when={data.state === 'errored'}>Error</Match>
  <Match when={data.state === 'ready'}>
    <For each={data()!}></For>
  </Match>
</Switch>
```

**Reply by ShueiYang** [reply]
Ok I see !  oh my I notice you are everywhere when it come to solving Solid.js xD

**Reply by snnsnn** [reply]
Yeah, trying to be helpful, thank you for noticing :)
